### PR TITLE
WIP: Add 5.0

### PR DESCRIPTION
**Pull request by aslicerh**
This is the start of trying to get things setup for 5.0. Currently it uses the MS bits but we will change it to use the Red Hat RPMs once they are available.

This is a WIP.

### CODE REVIEW

File: 5.0/build/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,60 @@
+FROM dotnet/dotnet-50-runtime-centos7
+# This image provides a .NET Core 5.0 environment you can use to run your .NET
+# applications.
+
+ENV PATH=/opt/app-root/src/.local/bin:/opt/app-root/src/bin:/opt/app-root/node_modules/.bin:${PATH} \
+    STI_SCRIPTS_PATH=/usr/libexec/s2i
+
+LABEL io.k8s.description="Platform for building and running .NET Core 5.0 applications" \
```
Comment (omajid) [inline 5.0/build/Dockerfile:8]: No "Core", just ".NET 5.0".

But "ASP.NET Core" is still "ASP.NET Core" :shrug:

File: 5.0/build/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      io.openshift.tags="builder,.net,dotnet,dotnetcore,rh-dotnet50"
```
Comment (omajid) [inline 5.0/build/Dockerfile:9]: we could/should probably leave out `dotnetcore` here too?

File: 5.0/build/test/asp-net-hello-world-envvar/src/app/app.csproj [hunk 3]
```diff
@@ -0,0 +1,25 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp5.0</TargetFramework>
```
Comment (omajid) [inline 5.0/build/test/asp-net-hello-world-envvar/src/app/app.csproj:4]: The TargetFramework for .NET 5.0 is just `net5.0`: https://github.com/dotnet/designs/blob/master/accepted/2020/net5/net5.md

File: 5.0/build/test/asp-net-hello-world-envvar/src/lib/lib.csproj [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
Comment (omajid) [inline 5.0/build/test/asp-net-hello-world-envvar/src/lib/lib.csproj:4]: `net5.0` is also the successor to `netstandard*`: https://github.com/dotnet/designs/blob/master/accepted/2020/net5/net5.md

File: 5.0/runtime/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+EXPOSE 8080
+
+ENV HOME=/opt/app-root \
+    PATH=/opt/app-root/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin \
+    DOTNET_APP_PATH=/opt/app-root/app \
+    DOTNET_DATA_PATH=/opt/app-root/data \
+    DOTNET_DEFAULT_CMD=default-cmd.sh \
+    DOTNET_CORE_VERSION=5.0 \
+    DOTNET_FRAMEWORK=netcoreapp5.0 \
```
Comment (omajid) [inline 5.0/runtime/Dockerfile:13]: This should be `net5.0`

File: 5.0/runtime/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+# applications.
+
+EXPOSE 8080
+
+ENV HOME=/opt/app-root \
+    PATH=/opt/app-root/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin \
+    DOTNET_APP_PATH=/opt/app-root/app \
+    DOTNET_DATA_PATH=/opt/app-root/data \
+    DOTNET_DEFAULT_CMD=default-cmd.sh \
+    DOTNET_CORE_VERSION=5.0 \
```
Comment (omajid) [inline 5.0/runtime/Dockerfile:12]: Do we want to change this var to `DOTNET_VERSION` now?

File: 5.0/build/Dockerfile.fedora [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      com.redhat.deployments-dir="/opt/app-root/src"
+
+# Switch to root for package installs
+USER 0
+
+# Copy the S2I scripts from the specific language image to $STI_SCRIPTS_PATH.
+COPY ./s2i/bin/ /usr/libexec/s2i
+
+# TODO: Change 3.1 to 5.0
+RUN INSTALL_PKGS="npm dotnet-sdk-3.1 rsync procps-ng findutils" && \
```
Comment (omajid) [inline 5.0/build/Dockerfile.fedora:29]: If you want to be adventurous, [the dotnet-sig/preview copr repo](https://copr.fedorainfracloud.org/coprs/g/dotnet-sig/dotnet-preview/) has preview 4 builds that you might be able to use here.

**Review by omajid — APPROVED** [review]
Nice! Thanks for getting this started!

File: 5.0/build/Dockerfile (hunk 1 above)
Comment (tmds) [reply 5.0/build/Dockerfile:8]: > But "ASP.NET Core" is still "ASP.NET Core" shrug

.NET 5 is the successor of .NET Framework 4, and .NET Core 3.0.
ASP.NET Core 5 isn't the successor of ASP.NET 4, so that's probably why "Core" stays in the name.

File: 5.0/build/Dockerfile.fedora (hunk 7 above)
Comment (tmds) [reply 5.0/build/Dockerfile.fedora:29]: :+1: it would be good to know how well tests work against the source-build .NET Core. This has revealed issues in the past.

File: 5.0/build/Dockerfile (hunk 2 above)
Comment (aslicerh) [reply 5.0/build/Dockerfile:9]: I left dotnetcore there because it has been a consistent tag across our whole product. I'm not stuck to that idea, but that was my thinking.

File: 5.0/runtime/Dockerfile (hunk 6 above)
Comment (aslicerh) [reply 5.0/runtime/Dockerfile:12]: I don't mind changing it. We just need to remember to change all the docs starting with 5.0 talking about the ENV vars.

@tmds Thoughts?

File: 5.0/build/Dockerfile (hunk 2 above)
Comment (tmds) [reply 5.0/build/Dockerfile:9]: I'm fine keeping it also so it shows as a "dotnetcore" match.

File: 5.0/runtime/Dockerfile (hunk 6 above)
Comment (tmds) [reply 5.0/runtime/Dockerfile:12]: If I had something automated using this environment variable, I'd prefer it continued to have the same name.
So I have a preference to keep it as is.

File: 5.0/build/Dockerfile (hunk 1 above)
Comment (aslicerh) [reply 5.0/build/Dockerfile]: Done.

File: 5.0/build/test/asp-net-hello-world-envvar/src/app/app.csproj (hunk 3 above)
Comment (aslicerh) [reply 5.0/build/test/asp-net-hello-world-envvar/src/app/app.csproj]: Done.

File: 5.0/build/test/asp-net-hello-world-envvar/src/lib/lib.csproj (hunk 4 above)
Comment (aslicerh) [reply 5.0/build/test/asp-net-hello-world-envvar/src/lib/lib.csproj]: Done.

File: 5.0/runtime/Dockerfile (hunk 5 above)
Comment (aslicerh) [reply 5.0/runtime/Dockerfile]: Done.

File: 5.0/build/Dockerfile.fedora (hunk 7 above)
Comment (aslicerh) [reply 5.0/build/Dockerfile.fedora:29]: I made those changes and the tests all passed. I didn't merge it here because the changes to the assemble/run scripts will make the other dockerfiles stop working (and I didn't want to make some kind of split logic in them for a temporary test).

But the change for it can be found in my personal branch [here](https://github.com/aslicerh/s2i-dotnetcore/tree/dev_50_use_copr)

**Comment by aslicerh** [comment]
@tmds 
I updated rh-nodejs to be version 14, and removed bower from the tests. Please take a look at the changes and let me know if all of that looks ok.

File: 5.0/build/Dockerfile (hunk 2 above)
Comment (omajid) [reply 5.0/build/Dockerfile:9]: Fair enough!

File: 5.0/build/Dockerfile.fedora (hunk 7 above)
Comment (omajid) [reply 5.0/build/Dockerfile.fedora:29]: Thanks for checking that source-build preview 4 works in this context!

File: 5.0/build/Dockerfile.fedora [hunk 8]
```diff
@@ -0,0 +1,58 @@
+FROM fedora/dotnet-50-runtime
+# This image provides a .NET Core 5.0 environment you can use to run your .NET
```
Comment (omajid) [inline 5.0/build/Dockerfile.fedora:2]: Can the "Core" here be removed?

File: 5.0/build/Dockerfile.fedora [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+# applications.
+
+ENV PATH=/opt/app-root/src/.local/bin:/opt/app-root/src/bin:/opt/app-root/node_modules/.bin:/opt/app-root/.dotnet/tools/:${PATH} \
+    STI_SCRIPTS_PATH=/usr/libexec/s2i
+
+LABEL io.k8s.description="Platform for building and running .NET Core 5.0 applications" \
```
Comment (omajid) [inline 5.0/build/Dockerfile.fedora:8]: Is this description here used by some search system? Is there a user-visible impact if "Core" is removed? I dont want to accidentally mislead users into thinking this is a "Core" release.

File: 5.0/build/README.md [hunk 10]
```diff
@@ -0,0 +1,211 @@
+.NET Core SDK image
```
Comment (omajid) [inline 5.0/build/README.md:1]: Can this "Core" also be removed?

File: 5.0/build/s2i/bin/usage [hunk 11]
```diff
@@ -0,0 +1,16 @@
+#!/bin/sh
+
+DISTRO=`cat /etc/*-release | grep ^ID= | grep -Po '".*?"' | tr -d '"'`
+NAMESPACE=dotnet
+
+cat <<EOF
+This is a S2I .Net core ${DISTRO} base image:
```
Comment (omajid) [inline 5.0/build/s2i/bin/usage:7]: Maybe we should say `.NET ${DISTRO}` instead of `.Net Core ${DISTRO}`?

File: 5.0/build/test/asp-net-hello-world-envvar2/src/app/NuGet.Config.s2i [hunk 12]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+  <packageSources>
+    <!--To inherit the global NuGet package sources remove the <clear/> line below -->
+    <clear />
+    <add key="nuget" value="https://api.nuget.org/v3/index.json" />
+    <add key="myget" value="https://www.myget.org/F/s2i-dotnetcore" />
```
Comment (omajid) [inline 5.0/build/test/asp-net-hello-world-envvar2/src/app/NuGet.Config.s2i:7]: Is the `myget` nuget repo required? It's used for non-public releases. Preview and GA releases should be available in the main nuget repo. I think we should leave it out if it's unneeded.

File: 5.0/runtime/root/usr/bin/cgroup-limits [hunk 13]
```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/python
+
+"""
+Script for parsing cgroup information
```
Comment (omajid) [inline 5.0/runtime/root/usr/bin/cgroup-limits:4]: Is this file still needed? Does it do anything that .NET 5 doesn't already do out of the box? .NET 5 is aware of cgroup v1 and cgroup v2.; this script seems support cgroup v1 only.

File: 5.0/build/test/asp-net-hello-world-envvar2/src/app/NuGet.Config.s2i (hunk 12 above)
Comment (tmds) [reply 5.0/build/test/asp-net-hello-world-envvar2/src/app/NuGet.Config.s2i:7]: This myget repo has a package that is used by one of the tests.

File: 5.0/runtime/root/usr/bin/cgroup-limits (hunk 13 above)
Comment (tmds) [reply 5.0/runtime/root/usr/bin/cgroup-limits:4]: This file never had any use for .NET.
We have been carrying it along for each release. I'm not aware of anyone depending it, so I don't mind if we remove it.

File: 5.0/build/Dockerfile.fedora (hunk 9 above)
Comment (tmds) [reply 5.0/build/Dockerfile.fedora:8]: +1 for moving Core from all places except tags.

Also, in descriptive text, like this, I think we should use ".NET 5" instead of ".NET 5.0".

File: 5.0/build/test/asp-net-hello-world-envvar2/src/app/NuGet.Config.s2i (hunk 12 above)
Comment (omajid) [reply 5.0/build/test/asp-net-hello-world-envvar2/src/app/NuGet.Config.s2i:7]: Doh, I missed that the url ends with `s2i-dotnetcore`. Thanks for clearing this up!

File: 5.0/build/Dockerfile.fedora (hunk 8 above)
Comment (aslicerh) [reply 5.0/build/Dockerfile.fedora]: Done.

File: 5.0/build/Dockerfile.fedora (hunk 9 above)
Comment (aslicerh) [reply 5.0/build/Dockerfile.fedora]: Done.

I didn't change it in the io.k8s.name because there will be times when 5.0 and 5.1 are side by side and will need to be differentiated between.

File: 5.0/build/README.md (hunk 10 above)
Comment (aslicerh) [reply 5.0/build/README.md]: Done.

File: 5.0/build/s2i/bin/usage (hunk 11 above)
Comment (aslicerh) [reply 5.0/build/s2i/bin/usage]: Done.

File: 5.0/runtime/root/usr/bin/cgroup-limits (hunk 13 above)
Comment (aslicerh) [reply 5.0/runtime/root/usr/bin/cgroup-limits]: I removed it.

File: 5.0/runtime/Dockerfile.rhel8 [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    DOTNET_DATA_PATH=/opt/app-root/data \
+    DOTNET_DEFAULT_CMD=default-cmd.sh \
+    DOTNET_CORE_VERSION=5.0 \
+    DOTNET_FRAMEWORK=net5.0 \
+# Microsoft's images set this to enable detecting when an app is running in a container.
+    DOTNET_RUNNING_IN_CONTAINER=true \
+    DOTNET_SSL_CERT_DIR=/opt/app-root/ssl_dir
+
+LABEL io.k8s.description="Platform for running .NET 5 applications" \
+      io.k8s.display-name=".NET 5.0" \
```
Comment (tmds) [inline 5.0/runtime/Dockerfile.rhel8:19]: I prefer to use ``.NET 5.0` also for `display-name`. There won't be a .NET 5.1 release. After .NET 5 comes the .NET 6 LTS release.

File: 5.0/build/Dockerfile [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+# Switch to root for package installs
+USER 0
+
+# Copy the S2I scripts from the specific language image to $STI_SCRIPTS_PATH.
+COPY ./s2i/bin/ /usr/libexec/s2i
+
+# TODO: Change 3.1 to 5.0
+RUN yum install -y centos-release-dotnet centos-release-scl-rh && \
+    INSTALL_PKGS="rh-nodejs14-npm rh-nodejs14-nodejs-nodemon rh-dotnet31-dotnet-sdk-3.1 rsync" && \
```
Comment (tmds) [inline 5.0/build/Dockerfile:30]: `rh-dotnet31-dotnet-sdk-3.1` refers to `3.1`.

File: 5.0/runtime/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    DOTNET_FRAMEWORK=net5.0 \
+# Microsoft's images set this to enable detecting when an app is running in a container.
+    DOTNET_RUNNING_IN_CONTAINER=true \
+    DOTNET_SSL_CERT_DIR=/opt/app-root/ssl_dir
+
+LABEL io.k8s.description="Platform for running .NET 5 applications" \
+      io.k8s.display-name=".NET 5.0" \
+      io.openshift.tags="runtime,.net,dotnet,dotnetcore,rh-dotnet31-runtime" \
```
Comment (tmds) [inline 5.0/runtime/Dockerfile:20]: `rh-dotnet31-runtime` refers to `3.1`.

File: 5.0/runtime/Dockerfile [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+      io.openshift.expose-services="8080:http" \
+      io.openshift.s2i.scripts-url=image:///usr/libexec/s2i \
+      io.s2i.scripts-url=image:///usr/libexec/s2i
+
+# Labels consumed by Red Hat build service
+LABEL name="dotnet/dotnet-31-runtime-centos7" \
```
Comment (tmds) [inline 5.0/runtime/Dockerfile:26]: `dotnet/dotnet-31-runtime-centos7` refers to `3.1`.

File: 5.0/runtime/Dockerfile [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+      com.redhat.component="rh-dotnet31-runtime-container" \
```
Comment (tmds) [inline 5.0/runtime/Dockerfile:27]: `rh-dotnet31-runtime-container` refers to `3.1`.

@aslicerh there are more like these: search for `3.1` and `31`.

File: 5.0/runtime/Dockerfile.rhel8 (hunk 14 above)
Comment (omajid) [reply 5.0/runtime/Dockerfile.rhel8:19]: I assume you mean `.NET 5`? and `5.0` is a typo?

File: 5.0/build/Dockerfile (hunk 15 above)
Comment (aslicerh) [reply 5.0/build/Dockerfile:30]: Yeah, there's a TODO on this one because we need to change it once the packages are available.

File: 5.0/runtime/Dockerfile.rhel8 (hunk 14 above)
Comment (tmds) [reply 5.0/runtime/Dockerfile.rhel8:19]: oops, yes I mean `.NET 5` instead of `.NET 5.0`.

File: 5.0/runtime/Dockerfile (hunk 16 above)
Comment (aslicerh) [reply 5.0/runtime/Dockerfile]: Done.

File: 5.0/runtime/Dockerfile (hunk 18 above)
Comment (aslicerh) [reply 5.0/runtime/Dockerfile]: Yeah, I found all those for the build dir, but must not of run one of my perl pies against the runtime.

They should be fixed now.

File: 5.0/runtime/Dockerfile (hunk 17 above)
Comment (aslicerh) [reply 5.0/runtime/Dockerfile]: Done.

File: 5.0/runtime/Dockerfile.rhel8 (hunk 14 above)
Comment (aslicerh) [reply 5.0/runtime/Dockerfile.rhel8]: Changed.

**Comment by tmds** [comment]
@aslicerh are there things to be done still before we can merge this?

Can you apply changes from https://github.com/redhat-developer/s2i-dotnetcore/pull/338 also to 5.0?

**Comment by aslicerh** [comment]
> @aslicerh are there things to be done still before we can merge this?
> 
> Can you apply changes from #338 also to 5.0?

Ok, I added the changes from #338.

As for merging, I wasn't sure if we wanted to merge it before the 5.0 GA version was available. Other than that, I believe that these changes should be sufficient as they are.

**Comment by tmds** [comment]
> Ok, I added the changes from #338.

Did you push these?

> As for merging, I wasn't sure if we wanted to merge it before the 5.0 GA version was available. Other than that, I believe that these changes should be sufficient as they are.

**Comment by azixaka** [comment]
Guys we can see the dotnet-50 image in RedHat catalogue is the preview-8. We thought the idea of having it would be to finally start getting GA releases on the same day as Microsoft release as you promised previously. However it's been 2 weeks since .net 5 is generally available but no image from RedHat yet. Could you please tell when will it be available?

**Comment by RheaAyase** [comment]
https://github.com/dotnet/core/issues/5596

**Comment by azixaka** [comment]
https://github.com/dotnet/source-build/issues/1913

**Comment by azixaka** [comment]
I think it's been tagged now? https://github.com/dotnet/core/tree/v5.0.0

Could you please tell if .net 5 images in redhat catalogue will appear soon as a result?

**Comment by omajid** [comment]
> I think it's been tagged now? https://github.com/dotnet/core/tree/v5.0.0

I *wish* that was the one we are waiting on. We are actually waiting on (and contributing to) https://github.com/dotnet/source-build/tags

**Comment by mml21** [comment]
Hello, do you have any updates on this? v5.0.100-SDK is available now.

Thank you

**Comment by azixaka** [comment]
The runtime is available too now. 
All eyes on RedHat now.

**Comment by tmds** [comment]
We'll add message here when .NET 5 containers are available.

We deliver .NET as fast as we can, unfortunately most of the delay is outside our team.

If you have a Red Hat subscription and you are impacted by this, consider contacting Red Hat about it. It won't do much for .NET 5 but it will help raise importance.

**Comment by RheaAyase** [comment]
Follow the parent issue please: https://github.com/dotnet/core/issues/5596
(Or did you expect us to work over night on a weekend? Just saying...)

**Comment by RheaAyase** [comment]
And indeed like Tom said - it is important to use Red Hat support as an official channel to request features, updates, or releases in timely manner. It helps us raise these issues not only within Red Hat but also with our partners (Microsoft.)

Please do :)

**Comment by aslicerh** [comment]
@azixaka @mml21 
5.0 images with the 5.0.100 SDK and runtime are now available in the RH repository.

**Comment by RheaAyase** [comment]
A side note that the documentation and official announcement are yet to be published.
